src/core/assertions.js

```diff
@@ -28,41 +28,48 @@ const RuleName = Object.freeze({
 })
 
 /**
- * Count object properties including nested objects ones.
- * If a property is an object, its key is ignored.
+ * Make an array of keys from an object.
+ * Use a dot notation
+ * Only empty object keys are taken into account
+ * For subobjects only keys are taken into account
  *
  * @example
- * Assertions.countNestedProperties({
+ * Assertions.buildObjectKeysArray({
  *     a: true,
  *     b: true,
  *     c: true,
  * })
- * // => 3
- * Assertions.countNestedProperties({
+ * // => ["a", "b", "c"]
+ * Assertions.buildObjectKeysArray({
  *     a: true,
  *     b: true,
  *     c: {
- *         a: true,
- *         b: true,
+ *         d: true,
+ *         e: {},
+ *         f: {
+ *              g: true
+ *          }
  *     },
  * })
- * // => 4 (c is ignored because it's a nested object)
+ * // =>  ["a", "b", "c.d", "c.e", "c.f.g"] (c and c.f are ignored as non empty nested objects)
  *
  * @param {Object} object
  * @return {number}
  */
-exports.countNestedProperties = (object) => {
-    let propertiesCount = 0
+exports.buildObjectKeysArray = (object, dottedObjectKeys = [], currentPath = '') => {
     Object.keys(object).forEach((key) => {
```

Comment:
With new conditions, if an object (or an array) is passed, you will get a result, otherwise an empty array is returned, and the exact assertions between spec and object will fail.  Maybe not the best way to do it, should need some improvement, but the recursive nature of  `objectKeysDeep`, will force us to detect if it's the first iteration and then validate the given input. Maybe should we check object correctness in `assertObjectMatchSpec` ?

Reply:
Good point! Yea, we could do this in `assertObjectMatchSpec`.